feat: Add OpenAI Responses API support to LLM class

enyst opened the pull request:
## Summary

This PR implements OpenAI Responses API support in the agent-sdk LLM class alongside the existing ChatCompletions API, as requested in issue #138.

We route `LLM.completion()` to the OpenAI Responses API (via LiteLLM) when the model supports it, convert `Responses` payloads into a ChatCompletions-like format, surface `assistant.message.reasoning_content`, and record `usage.reasoning_tokens`. Telemetry persists raw Responses payloads for audit.

## Changes

### Core Implementation
- **Added `supports_responses_api` field** to `ModelFeatures` dataclass
- **Created `responses_converter.py`** with utilities to convert between ChatCompletions and Responses API formats
- **Implemented `responses()` method** in LLM class using `litellm.responses()`

### Features
- **Format conversion**: between message formats and string input/output
- **Model compatibility detection**: detects which models support Responses API
- **Backward compatibility**: Existing `completion()` method remains unchanged

## Supported Models

The implementation supports Responses API for:
- **OpenAI reasoning models**: GPT-5, GPT-5-mini, o1-preview, o1-mini, o3-mini.

## Usage Example

```python
from openhands.sdk.llm import LLM

# For reasoning models that support Responses API
llm = LLM(model="o1-preview")
if llm.is_responses_api_supported():
    response = llm.responses(
        input="Solve this step by step: What is 2+2?"
    )
    print(response.choices[0].message.content)

# Traditional ChatCompletions API still works
response = llm.completion(
    messages=[{"role": "user", "content": "What is 2+2?"}]
)
```

## Testing Results

All tests pass:
- ✅ 7/7 converter utility tests
- ✅ 22/22 model feature tests  
- ✅ 10/10 LLM responses method tests
- ✅ All existing tests continue to pass

Fixes #138

@enyst can click here to [continue refining the PR](https://app.all-hands.dev/conversations/ff4b69c0b2f540be8fe460aa5f2425e0)


<!-- AGENT_SERVER_IMAGES_START -->
---
**Agent Server images for this PR**

• **GHCR package:** https://github.com/All-Hands-AI/agent-sdk/pkgs/container/agent-server

**Variants & Base Images**
| Variant | Base Image | Docs / Tags |
|---|---|---|
| golang | `golang:1.21-bookworm` | [Link](https://hub.docker.com/_/golang:1.21-bookworm) |
| java | `eclipse-temurin:17-jdk` | [Link](https://hub.docker.com/_/eclipse-temurin:17-jdk) |
| python | `nikolaik/python-nodejs:python3.12-nodejs22` | [Link](https://hub.docker.com/_/nikolaik/python-nodejs:python3.12-nodejs22) |


**Pull (multi-arch manifest)**
```bash
docker pull ghcr.io/all-hands-ai/agent-server:68257dd-python
```

**Run**
```bash
docker run -it --rm \
  -p 8000:8000 \
  --name agent-server-68257dd-python \
  ghcr.io/all-hands-ai/agent-server:68257dd-python
```

**All tags pushed for this build**
```
ghcr.io/all-hands-ai/agent-server:68257dd-golang
ghcr.io/all-hands-ai/agent-server:v1.0.0_golang_tag_1.21-bookworm_golang
ghcr.io/all-hands-ai/agent-server:68257dd-java
ghcr.io/all-hands-ai/agent-server:v1.0.0_eclipse-temurin_tag_17-jdk_java
ghcr.io/all-hands-ai/agent-server:68257dd-python
ghcr.io/all-hands-ai/agent-server:v1.0.0_nikolaik_s_python-nodejs_tag_python3.12-nodejs22_python
```

_The `68257dd` tag is a multi-arch manifest (amd64/arm64); your client pulls the right arch automatically._
<!-- AGENT_SERVER_IMAGES_END -->

enyst commented on openhands/sdk/llm/utils/model_features.py:
```diff
@@ -124,6 +125,36 @@ class ModelFeatures:
+    "claude-opus-4*",
+    # Google Gemini models (supported via LiteLLM bridge)
+    "gemini-1.5-pro*",
+    "gemini-2.0-flash*",
```
@openhands Double check every model in the list of RESPONSES_API_PATTERNS against litellm's documentation or code for responses API. For example, gemini 1.5 pro does NOT have reasoning. Don't hallucinate, check. If you can't find a model, just remove it from this list.

enyst commented on openhands/sdk/llm/utils/responses_converter.py:
```diff
@@ -0,0 +1,143 @@
+        usage = responses_result.usage
+        if hasattr(usage, "input_tokens"):
+            response["usage"]["prompt_tokens"] = usage.input_tokens
+        if hasattr(usage, "output_tokens"):
```
@openhands Clone litellm repo and verify this: if a response from Responses API has 'usage', does the usage object ALWAYS have properties used in this code, e.g. input_tokens, output_tokens, completion_... etc, or not?

If always, then clean up the relevant hasattr() conditional(s), clean up the code!

enyst commented:
@openhands fix the failing llm tests

enyst commented:
@openhands Read this PR and understand what it aims to achieve.

Then read this file and copy it in your workspace, in agent-sdk examples/ for testing:

https://raw.githubusercontent.com/All-Hands-AI/agent-sdk/1b06c01eac7f9170bd92591aa5f03e9bfa988b78/examples/10_reasoning_debug.py

Check the agent-sdk README.md file for how to run these example files easily, same applies to the new file.

Add gpt-5-mini-2025-08-07 or another gpt-5-mini model that exists on the litellm proxy, to the new examples file in the list of default models. Comment out all the other models.

Goal: run the debug file to verify that it works to show reasoning for GPT-5-mini*, and fix any issues.

When you are done, make a PR to this PR, so we can see.

xingyaoww commented on openhands/sdk/llm/llm.py:
```diff
@@ -583,6 +679,66 @@ def _all_choices(
+            out.pop("top_p", None)
+
+        # Remove parameters not supported by Responses API
+        out.pop("tools", None)
```
Are we sure we want to remove tools?

it does seems to be supported
https://platform.openai.com/docs/api-reference/responses/create#responses_create-tools

enyst replied:
Good point, and it definitely needs it! I understand from docs that it can do reasoning between tool calls

enyst replied:
Fixed!

xingyaoww reviewed (commented):
Some GPT-5 suggestion into how we might simplify the `llm.py` code

```
from typing import Any, Literal, Union, Annotated, cast
from pydantic import BaseModel, Field, ConfigDict
# ... keep your existing imports

CallKind = Literal["chat", "responses"]

# ---------------------------
# Discriminated union context
# ---------------------------

class _BaseCtxModel(BaseModel):
    model_config = ConfigDict(extra="forbid")
    call_kwargs: dict[str, Any]
    log_ctx: dict[str, Any]
    tools: list[ChatCompletionToolParam] = Field(default_factory=list)

class ChatCtx(_BaseCtxModel):
    kind: Literal["chat"]
    messages: list[dict[str, Any]]
    # internal to support mocked tool-calls
    nonfn_msgs: list[dict[str, Any]] | None = None
    raw_resp_needed: bool = False

class ResponsesCtx(_BaseCtxModel):
    kind: Literal["responses"]
    input: str

ReqCtx = Annotated[Union[ChatCtx, ResponsesCtx], Field(discriminator="kind")]

# ---------------------------
# Public wrappers: unchanged API
# ---------------------------

def completion(
    self,
    messages: list[dict[str, Any]] | list[Message],
    tools: list[ChatCompletionToolParam] | None = None,
    return_metrics: bool = False,
    **kwargs,
) -> ModelResponse:
    if kwargs.get("stream", False):
        raise ValueError("Streaming is not supported")

    # Serialize Message objects if present
    if messages and isinstance(messages[0], Message):
        messages = self.format_messages_for_llm(cast(list[Message], messages))
    else:
        messages = cast(list[dict[str, Any]], messages)

    return self._unified_request(
        kind="chat",
        messages=messages,
        tools=tools or [],
        **kwargs,
    )


def responses(
    self,
    messages: list[dict[str, Any]] | list[Message] | str | None = None,
    input: str | None = None,
    **kwargs,
) -> ModelResponse:
    if not self.is_responses_api_supported():
        raise ValueError(
            f"Model {self.model} does not support the Responses API. Use completion()."
        )
    if kwargs.get("stream", False):
        raise ValueError("Streaming is not supported in responses() method")

    # Determine input text
    if input is None:
        if messages is None:
            raise ValueError("Either 'messages' or 'input' must be provided")
        if isinstance(messages, str):
            input = messages
        else:
            input = messages_to_responses_input(messages)

    return self._unified_request(kind="responses", input=input, **kwargs)

# ---------------------------
# Unified engine
# ---------------------------

def _unified_request(
    self,
    *,
    kind: CallKind,
    messages: list[dict[str, Any]] | None = None,
    input: str | None = None,
    tools: list[ChatCompletionToolParam] | None = None,
    **kwargs,
) -> ModelResponse:
    assert self._telemetry is not None

    # 1) Build validated context (Pydantic will enforce required fields by kind)
    ctx: ReqCtx = self._build_ctx(kind=kind, messages=messages, input=input, tools=tools or [], opts=kwargs)

    # 2) Telemetry (request)
    self._telemetry.on_request(log_ctx=ctx.log_ctx)

    # 3) Retry wrapper
    @self.retry_decorator(
        num_retries=self.num_retries,
        retry_exceptions=LLM_RETRY_EXCEPTIONS,
        retry_min_wait=self.retry_min_wait,
        retry_max_wait=self.retry_max_wait,
        retry_multiplier=self.retry_multiplier,
        retry_listener=self.retry_listener,
    )
    def _one_attempt() -> ModelResponse:
        # Transport
        resp = self._transport_dispatch(ctx)
        # Post-process (mocked tool-calls)
        resp = self._postprocess_dispatch(ctx, resp)
        # Response telemetry + invariant
        self._telemetry.on_response(resp)
        if not resp.get("choices") or len(resp["choices"]) < 1:
            raise LLMNoResponseError("Response choices is less than 1. Response: " + str(resp))
        return resp

    try:
        return _one_attempt()
    except Exception as e:
        self._telemetry.on_error(e)
        raise

# ---------------------------
# Build + normalize once
# ---------------------------

def _build_ctx(
    self,
    *,
    kind: CallKind,
    messages: list[dict[str, Any]] | None,
    input: str | None,
    tools: list[ChatCompletionToolParam],
    opts: dict[str, Any],
) -> ReqCtx:
    use_native_fc = self.is_function_calling_active()

    nonfn_msgs: list[dict[str, Any]] | None = None
    raw_resp_needed = False

    # Mock tool-calls only for chat when native FC is off
    if kind == "chat" and tools and not use_native_fc:
        nonfn_msgs = copy.deepcopy(messages or [])
        messages, opts = self._pre_request_prompt_mock(messages or [], tools, opts)
        raw_resp_needed = True

    has_tools = bool(tools) and (kind == "chat") and use_native_fc
    call_kwargs = self._normalize_kwargs(kind, opts, has_tools=has_tools)

    # Build log context (what we write is exactly what we send)
    log_ctx: dict[str, Any] = {
        "kwargs": {k: v for k, v in call_kwargs.items()},
        "context_window": self.max_input_tokens,
    }
    if kind == "chat":
        log_ctx["messages"] = (messages or [])[:]
        log_ctx["tools"] = tools if has_tools else None
        if nonfn_msgs is not None:
            log_ctx["raw_messages"] = nonfn_msgs
        return ChatCtx(
            kind="chat",
            messages=messages or [],
            nonfn_msgs=nonfn_msgs,
            raw_resp_needed=raw_resp_needed,
            call_kwargs=call_kwargs,
            log_ctx=log_ctx,
            tools=tools,
        )
    else:
        log_ctx["input"] = input or ""
        return ResponsesCtx(
            kind="responses",
            input=input or "",
            call_kwargs=call_kwargs,
            log_ctx=log_ctx,
            tools=tools,
        )

def _normalize_kwargs(self, kind: CallKind, opts: dict, *, has_tools: bool) -> dict:
    out = dict(opts)

    # Sampling defaults
    if self.top_k is not None: out.setdefault("top_k", self.top_k)
    if self.top_p is not None: out.setdefault("top_p", self.top_p)
    if self.temperature is not None: out.setdefault("temperature", self.temperature)

    # Max tokens (chat vs responses)
    if kind == "chat":
        if self.max_output_tokens is not None:
            out.setdefault("max_completion_tokens", self.max_output_tokens)
        if self.model.startswith("azure") and "max_completion_tokens" in out:
            out["max_tokens"] = out.pop("max_completion_tokens")
    else:
        if self.max_output_tokens is not None:
            out.setdefault("max_output_tokens", self.max_output_tokens)

    # Reasoning-model quirks
    if get_features(self.model).supports_reasoning_effort:
        if self.reasoning_effort is not None:
            out["reasoning_effort"] = self.reasoning_effort
        out.pop("temperature", None)
        out.pop("top_p", None)
        if "gemini-2.5-pro" in self.model and out.get("reasoning_effort") in (None, "none"):
            out["reasoning_effort"] = "low"

    # Mistral / Gemini safety
    if self.safety_settings:
        ml = self.model.lower()
        if "mistral" in ml or "gemini" in ml:
            out["safety_settings"] = self.safety_settings

    # Tools only when chat+native FC
    if kind != "chat" or not has_tools:
        out.pop("tools", None)
        out.pop("tool_choice", None)

    # Responses API doesn't support stop/tools
    if kind == "responses":
        out.pop("stop", None)

    # Only keep extra_body for litellm_proxy
    if "litellm_proxy" not in self.model:
        out.pop("extra_body", None)

    return out

# ---------------------------
# Transport + postprocess
# ---------------------------

def _transport_dispatch(self, ctx: ReqCtx) -> ModelResponse:
    with self._litellm_modify_params_ctx(self.modify_params):
        with warnings.catch_warnings():
            warnings.filterwarnings("ignore", category=DeprecationWarning, module="httpx.*")
            warnings.filterwarnings("ignore", message=r".*content=.*upload.*", category=DeprecationWarning)

            if ctx.kind == "chat":
                ret = litellm_completion(
                    model=self.model,
                    api_key=self.api_key.get_secret_value() if self.api_key else None,
                    base_url=self.base_url,
                    api_version=self.api_version,
                    timeout=self.timeout,
                    drop_params=self.drop_params,
                    seed=self.seed,
                    messages=ctx.messages,               # type: ignore[attr-defined]
                    **ctx.call_kwargs,
                )
                assert isinstance(ret, ModelResponse)
                return ret
            else:
                raw = litellm_responses(
                    model=self.model,
                    api_key=self.api_key.get_secret_value() if self.api_key else None,
                    base_url=self.base_url,
                    api_version=self.api_version,
                    timeout=self.timeout,
                    input=ctx.input,                      # type: ignore[attr-defined]
                    **ctx.call_kwargs,
                )
                return responses_to_completion_format(raw)

def _postprocess_dispatch(self, ctx: ReqCtx, resp: ModelResponse) -> ModelResponse:
    # tool-mocking only applies to chat when we mocked pre-request
    if ctx.kind == "chat" and getattr(ctx, "raw_resp_needed", False):  # type: ignore[attr-defined]
        raw_resp = copy.deepcopy(resp)
        resp = self._post_response_prompt_mock(
            resp,
            nonfncall_msgs=getattr(ctx, "nonfn_msgs") or [],              # type: ignore[attr-defined]
            tools=ctx.tools,
        )
        # keep raw vs converted for telemetry parity
        self._telemetry.on_response(resp, raw_resp=raw_resp)
        return resp
    return resp
```

enyst commented:
@openhands fix this test 

FAILED tests/sdk/llm/test_llm_responses.py::test_responses_method_parameter_normalization - AssertionError: assert 'tools' not in {'model': 'o1-preview', 'api_key': None, 'base_url': None, 'api_version': None, 'timeout': None, 'input': 'Test input', 'tools': [{'type': 'function', 'function': {'name': 'test'}}], 'max_output_tokens': 1000, 'reasoning_effort': 'medium'}
================== 1 failed, 499 passed, 2 warnings in 22.24s ==================
Error: Process completed with exit code 1.

enyst commented:
### Quick notes from experiments with
 `litellm_proxy/openai/gpt-5-mini-2025-08-07`

(I need to understand how we can take advantage of this _stateful_ API.)

#### Input format:
```
"kwargs": {
  "max_output_tokens": 128000,
  "reasoning_effort": "high",
  "reasoning": {
    "effort": "high",
    "summary": "detailed"
  }
}
- OpenAI: summary - "can be useful for debugging and understanding the model's reasoning process. One of auto, concise, or detailed."
- NOTE: apparently our litellm proxy doesn't support `concise`.
```

#### Responses API - specific `input` field:
- text
- image
- file inputs
```
"input": "Solve this carefully: 78*964 + 17."
```

Responses API stateful-specific field (we don't use it atm): **`previous_response_id`**


**Response:** converted to ChatCompletion API response, populates `reasoning_content` and `reasoning_tokens`.

**Note:** has a 'truncation' option to not fail with 400 on context window size exceeded:
- https://platform.openai.com/docs/api-reference/responses/create#responses-create-truncation

**Note:** has endpoints to return response by response id, cancel (streaming), and to delete a response
- https://platform.openai.com/docs/api-reference/responses/create#responses_create-tools

Response:
- has a [`conversation`](https://platform.openai.com/docs/api-reference/responses/object#responses/object-conversation)
- [output](https://platform.openai.com/docs/api-reference/responses/object#responses/object-output)
  - reasoning
    - "Be sure to include these items in your input to the Responses API for subsequent turns of a conversation if you are manually [managing context](https://platform.openai.com/docs/guides/conversation-state)."
    - summary `{"type": "summary_text", "summary": "text"}`
        - A summary of the reasoning output from the model so far.
    - content `{"type": "reasoning_text", "text": "text"}`
        - Reasoning text output from the model.

Question: does the API return multiple `summary` fields?
- how does it work with parallel tool calls?

enyst commented:
## Update: End-to-end Responses API via LiteLLM proxy (gpt-5-mini-2025-08-07)

- Verified that `LLM.completion()` auto-routes to the OpenAI Responses API for models with `supports_responses_api=True` and that `assistant.message.reasoning_content` plus `usage.reasoning_tokens` are surfaced end-to-end via LiteLLM proxy.
- Example script updated to focus solely on the proxy model `litellm_proxy/openai/gpt-5-mini-2025-08-07` and to persist audit logs (including raw Responses payloads) for inspection.

**How to run**
- Env:
  - `LITELLM_API_KEY` (required)
  - `LITELLM_BASE_URL` (defaults to `https://llm-proxy.eval.all-hands.dev`)
  - Optional: `REASONING_LOG_DIR` (default `logs/reasoning`)
- Command:
  - `PYTHONPATH=. python -u examples_local/10_reasoning_debug.py`

**Results (proxy: llm-proxy.eval.all-hands.dev)**
- `completion() -> Responses` path: reasoning_content present; `reasoning_tokens` observed ≈ 960.
- Direct `responses()` path: reasoning_content present; `reasoning_tokens` observed across runs ≈ 640–896.
- Audit logs written under `logs/reasoning/` (per-call JSON files include `response`, `raw_response`, `usage_summary` with `reasoning_tokens`).

**Main files (one sentence each)**
- `openhands/sdk/llm/llm.py` — Routes `completion()` to the Responses API for models that support it, normalizes request params for reasoning models, and preserves reasoning fields during any prompt-mock conversion.
- `openhands/sdk/llm/utils/responses_converter.py` — Converts Responses objects into a ChatCompletions-like `ModelResponse`, surfacing `assistant.message.reasoning_content` and mapping `usage.output_tokens_details.reasoning_tokens` to `usage.completion_tokens_details.reasoning_tokens`.
- `openhands/sdk/llm/utils/telemetry.py` — Persists structured JSON logs per call, including `response`, `raw_response`, latency, cost, and a compact `usage_summary` (prompt/completion/reasoning tokens).
- `examples_local/10_reasoning_debug.py` — Minimal, single-model example using the LiteLLM proxy to validate `completion()`→Responses routing and the direct `responses()` path, saving audit artifacts for review.

**Notes**
- The example now exclusively targets `litellm_proxy/openai/gpt-5-mini-2025-08-07` to keep validation deterministic.
- Co-authored-by on the latest example change: `openhands <openhands@all-hands.dev>`.

enyst commented:
GPT-5-mini, running `examples/5_use_llm_registry.py`
<img width="797" height="338" alt="image" src="https://github.com/user-attachments/assets/83afa717-58a2-4bf5-8e94-8d49966d2137" />

enyst commented:
Result of `/codereview-roasted`:

> Taste Rating:
> 🟡 Acceptable – Solves real problems (Responses API tooling, backwards mapping), but there are areas where the implementation could be cleaner, less error-prone, and more principled.
> 
> [CRITICAL ISSUES] (Must fix)
> - [openhands/sdk/agent/agent/agent.py, ~152–160] Complexity/Design: Test-specific behavior in production code. The step() method conditionally forces force_chat_completions=True when self.llm is not an instance of LLM (trying to infer tests/mocks). This is brittle and violates separation of concerns. It risks breaking legitimate custom LLM wrappers that aren’t the exact LLM class. Remove this logic and handle forcing in tests, as you indicated has been done.
> - [openhands/sdk/llm/llm.py, _normalize_responses_kwargs(), ~776–833] Data correctness: Tool conversion emits flattened tools with description=None and parameters=None. Sending explicit None values to providers is risky and can cause spurious 400s. Only include keys that have values. Don’t add description or parameters if absent.
> - [openhands/sdk/llm/llm.py, _normalize_responses_kwargs(), multiple spots] Observability/debuggability: Broad try/except with silent pass around tool conversion and tool_choice mapping will hide real bugs and make production issues hard to diagnose. At minimum, log/debug when conversion fails and include shape of the offending tool. Silent failure here will be a nightmare to debug.
> 
> [IMPROVEMENT OPPORTUNITIES] (Should fix)
> - [openhands/sdk/llm/llm.py, _normalize_responses_kwargs(), ~769–839] Complexity: The conversion block for tools tries to handle several input types (our Tool class with to_responses_tool, ChatCompletionToolParam-like objects, dicts, miscellaneous). This leads to special-case-heavy code. A simpler approach:
>   - Normalize inputs early to a single internal representation (dicts) at API boundary.
>   - Then perform a single, clean conversion to flattened Responses tools.
>   - Eliminate hasattr gymnastics and reduce branches. Good taste: reduce the number of cases by normalizing vs branching on every possible input.
> - [openhands/sdk/llm/utils/responses_converter.py, responses_to_completion_format(), ~80–156] Single-responsibility/clarity: The converter is doing three distinct jobs (content extraction, tool-call mapping, reasoning mapping) with lots of defensive programming. Extract helpers:
>   - _extract_message_content(item)
>   - _extract_tool_call(item)
>   - _extract_reasoning_content(item)
>   These helpers should return clean values or None. The main loop becomes a simple “collect from helpers” pass, reducing nesting and improving readability.
> - [openhands/sdk/llm/llm.py, responses() and _transport_responses_call()] Pragmatism: The addition of both “reasoning_effort” and a defaulted “reasoning” block (with summary: detailed) is useful, but it should be guarded/justified per provider. Consider:
>   - Only set “reasoning” if the model is known to honor it (OpenAI reasoning models), and not for others; even with litellm this is safer. You’ve already checked supports_reasoning_effort; add a short code comment why we’re setting “reasoning” and the intended cross-provider behavior.
> - [openhands/sdk/tool/tool.py, to_responses_tool()] Data cleanliness: Same as above — do not include keys with None. Only set description/parameters if present. Many APIs treat null as different from omitting the key.
> - [openhands/sdk/llm/llm.py, _transport_responses_call()] Edge cases: api_base (base_url) handling is subtle for proxies vs direct providers. Consider asserting/validating that api_base is present (or defaulted) when using litellm.responses() to avoid hard-to-debug misconfigurations.
> 
> [STYLE NOTES] (Minor)
> - [openhands/sdk/llm/llm.py, multiple] Logging: Use debug logs when conversion branches are chosen, especially in the tool conversion path, to make support easier. Today, broad except with pass hides valuable signals.
> - [openhands/sdk/llm/utils/responses_converter.py, ~146–205] Data conversion approach is reasonable, but providing a single “schema version” comment at the top explaining expected Responses items (message, function_call, reasoning) would help future maintainers.
> 
> VERDICT:
> ✅ Worth merging – It addresses a real issue (OpenAI Responses API tool shape) and maintains the agent flow by mapping tool calls back to Chat Completions format. With a few targeted fixes, it will be robust and maintainable.
> 
> KEY INSIGHT:
> The core problem was real (OpenAI Responses tool shape vs Chat Completions). The solution works, but the tool conversion layer and test accommodations leaked into production code in one spot (Agent.step). Consolidate conversion logic (normalize then convert), avoid silent failures, and keep test-specific forcing out of the runtime path. This will simplify the code and reduce the risk of future regressions without sacrificing functionality.

enyst commented:
Refactor: unified request engine for completion()/responses()

I implemented the simplification we discussed. Key points:
- Added a single internal method `_unified_request(kind)` that handles both chat and responses flows.
- Centralizes: parameter normalization, telemetry (request/response), retries, transport dispatch, and postprocess.
- For Responses API: keeps tools supported (converted via our Tool.to_responses / dict-flatten), removes unsupported stop, applies reasoning_effort rules, and converts back to Chat Completions for compatibility.
- For Chat: preserves prompt-mocked tool-calls when native FC is off; otherwise passes native tools when active.
- Public API remains unchanged. Tests and ruff checks pass locally.

Co-authored-by: OpenHands-GPT-5 <openhands@all-hands.dev>

xingyaoww reviewed (commented):
Hey @enyst , your plan sounds good to me!

Just a few comments! (We'd better be strict now than finding ourselves looking at sonnet-generated Spaghetti code in the future)

---

I inteneded to send this earlier but somehow i didn't "Submit review" so it stays pending 😓 

Some GPT-5 suggestion into how we might simplify the `llm.py` code

```
from typing import Any, Literal, Union, Annotated, cast
from pydantic import BaseModel, Field, ConfigDict
# ... keep your existing imports

CallKind = Literal["chat", "responses"]

# ---------------------------
# Discriminated union context
# ---------------------------

class _BaseCtxModel(BaseModel):
    model_config = ConfigDict(extra="forbid")
    call_kwargs: dict[str, Any]
    log_ctx: dict[str, Any]
    tools: list[ChatCompletionToolParam] = Field(default_factory=list)

class ChatCtx(_BaseCtxModel):
    kind: Literal["chat"]
    messages: list[dict[str, Any]]
    # internal to support mocked tool-calls
    nonfn_msgs: list[dict[str, Any]] | None = None
    raw_resp_needed: bool = False

class ResponsesCtx(_BaseCtxModel):
    kind: Literal["responses"]
    input: str

ReqCtx = Annotated[Union[ChatCtx, ResponsesCtx], Field(discriminator="kind")]

# ---------------------------
# Public wrappers: unchanged API
# ---------------------------

def completion(
    self,
    messages: list[dict[str, Any]] | list[Message],
    tools: list[ChatCompletionToolParam] | None = None,
    return_metrics: bool = False,
    **kwargs,
) -> ModelResponse:
    if kwargs.get("stream", False):
        raise ValueError("Streaming is not supported")

    # Serialize Message objects if present
    if messages and isinstance(messages[0], Message):
        messages = self.format_messages_for_llm(cast(list[Message], messages))
    else:
        messages = cast(list[dict[str, Any]], messages)

    return self._unified_request(
        kind="chat",
        messages=messages,
        tools=tools or [],
        **kwargs,
    )


def responses(
    self,
    messages: list[dict[str, Any]] | list[Message] | str | None = None,
    input: str | None = None,
    **kwargs,
) -> ModelResponse:
    if not self.is_responses_api_supported():
        raise ValueError(
            f"Model {self.model} does not support the Responses API. Use completion()."
        )
    if kwargs.get("stream", False):
        raise ValueError("Streaming is not supported in responses() method")

    # Determine input text
    if input is None:
        if messages is None:
            raise ValueError("Either 'messages' or 'input' must be provided")
        if isinstance(messages, str):
            input = messages
        else:
            input = messages_to_responses_input(messages)

    return self._unified_request(kind="responses", input=input, **kwargs)

# ---------------------------
# Unified engine
# ---------------------------

def _unified_request(
    self,
    *,
    kind: CallKind,
    messages: list[dict[str, Any]] | None = None,
    input: str | None = None,
    tools: list[ChatCompletionToolParam] | None = None,
    **kwargs,
) -> ModelResponse:
    assert self._telemetry is not None

    # 1) Build validated context (Pydantic will enforce required fields by kind)
    ctx: ReqCtx = self._build_ctx(kind=kind, messages=messages, input=input, tools=tools or [], opts=kwargs)

    # 2) Telemetry (request)
    self._telemetry.on_request(log_ctx=ctx.log_ctx)

    # 3) Retry wrapper
    @self.retry_decorator(
        num_retries=self.num_retries,
        retry_exceptions=LLM_RETRY_EXCEPTIONS,
        retry_min_wait=self.retry_min_wait,
        retry_max_wait=self.retry_max_wait,
        retry_multiplier=self.retry_multiplier,
        retry_listener=self.retry_listener,
    )
    def _one_attempt() -> ModelResponse:
        # Transport
        resp = self._transport_dispatch(ctx)
        # Post-process (mocked tool-calls)
        resp = self._postprocess_dispatch(ctx, resp)
        # Response telemetry + invariant
        self._telemetry.on_response(resp)
        if not resp.get("choices") or len(resp["choices"]) < 1:
            raise LLMNoResponseError("Response choices is less than 1. Response: " + str(resp))
        return resp

    try:
        return _one_attempt()
    except Exception as e:
        self._telemetry.on_error(e)
        raise

# ---------------------------
# Build + normalize once
# ---------------------------

def _build_ctx(
    self,
    *,
    kind: CallKind,
    messages: list[dict[str, Any]] | None,
    input: str | None,
    tools: list[ChatCompletionToolParam],
    opts: dict[str, Any],
) -> ReqCtx:
    use_native_fc = self.is_function_calling_active()

    nonfn_msgs: list[dict[str, Any]] | None = None
    raw_resp_needed = False

    # Mock tool-calls only for chat when native FC is off
    if kind == "chat" and tools and not use_native_fc:
        nonfn_msgs = copy.deepcopy(messages or [])
        messages, opts = self._pre_request_prompt_mock(messages or [], tools, opts)
        raw_resp_needed = True

    has_tools = bool(tools) and (kind == "chat") and use_native_fc
    call_kwargs = self._normalize_kwargs(kind, opts, has_tools=has_tools)

    # Build log context (what we write is exactly what we send)
    log_ctx: dict[str, Any] = {
        "kwargs": {k: v for k, v in call_kwargs.items()},
        "context_window": self.max_input_tokens,
    }
    if kind == "chat":
        log_ctx["messages"] = (messages or [])[:]
        log_ctx["tools"] = tools if has_tools else None
        if nonfn_msgs is not None:
            log_ctx["raw_messages"] = nonfn_msgs
        return ChatCtx(
            kind="chat",
            messages=messages or [],
            nonfn_msgs=nonfn_msgs,
            raw_resp_needed=raw_resp_needed,
            call_kwargs=call_kwargs,
            log_ctx=log_ctx,
            tools=tools,
        )
    else:
        log_ctx["input"] = input or ""
        return ResponsesCtx(
            kind="responses",
            input=input or "",
            call_kwargs=call_kwargs,
            log_ctx=log_ctx,
            tools=tools,
        )

def _normalize_kwargs(self, kind: CallKind, opts: dict, *, has_tools: bool) -> dict:
    out = dict(opts)

    # Sampling defaults
    if self.top_k is not None: out.setdefault("top_k", self.top_k)
    if self.top_p is not None: out.setdefault("top_p", self.top_p)
    if self.temperature is not None: out.setdefault("temperature", self.temperature)

    # Max tokens (chat vs responses)
    if kind == "chat":
        if self.max_output_tokens is not None:
            out.setdefault("max_completion_tokens", self.max_output_tokens)
        if self.model.startswith("azure") and "max_completion_tokens" in out:
            out["max_tokens"] = out.pop("max_completion_tokens")
    else:
        if self.max_output_tokens is not None:
            out.setdefault("max_output_tokens", self.max_output_tokens)

    # Reasoning-model quirks
    if get_features(self.model).supports_reasoning_effort:
        if self.reasoning_effort is not None:
            out["reasoning_effort"] = self.reasoning_effort
        out.pop("temperature", None)
        out.pop("top_p", None)
        if "gemini-2.5-pro" in self.model and out.get("reasoning_effort") in (None, "none"):
            out["reasoning_effort"] = "low"

    # Mistral / Gemini safety
    if self.safety_settings:
        ml = self.model.lower()
        if "mistral" in ml or "gemini" in ml:
            out["safety_settings"] = self.safety_settings

    # Tools only when chat+native FC
    if kind != "chat" or not has_tools:
        out.pop("tools", None)
        out.pop("tool_choice", None)

    # Responses API doesn't support stop/tools
    if kind == "responses":
        out.pop("stop", None)

    # Only keep extra_body for litellm_proxy
    if "litellm_proxy" not in self.model:
        out.pop("extra_body", None)

    return out

# ---------------------------
# Transport + postprocess
# ---------------------------

def _transport_dispatch(self, ctx: ReqCtx) -> ModelResponse:
    with self._litellm_modify_params_ctx(self.modify_params):
        with warnings.catch_warnings():
            warnings.filterwarnings("ignore", category=DeprecationWarning, module="httpx.*")
            warnings.filterwarnings("ignore", message=r".*content=.*upload.*", category=DeprecationWarning)

            if ctx.kind == "chat":
                ret = litellm_completion(
                    model=self.model,
                    api_key=self.api_key.get_secret_value() if self.api_key else None,
                    base_url=self.base_url,
                    api_version=self.api_version,
                    timeout=self.timeout,
                    drop_params=self.drop_params,
                    seed=self.seed,
                    messages=ctx.messages,               # type: ignore[attr-defined]
                    **ctx.call_kwargs,
                )
                assert isinstance(ret, ModelResponse)
                return ret
            else:
                raw = litellm_responses(
                    model=self.model,
                    api_key=self.api_key.get_secret_value() if self.api_key else None,
                    base_url=self.base_url,
                    api_version=self.api_version,
                    timeout=self.timeout,
                    input=ctx.input,                      # type: ignore[attr-defined]
                    **ctx.call_kwargs,
                )
                return responses_to_completion_format(raw)

def _postprocess_dispatch(self, ctx: ReqCtx, resp: ModelResponse) -> ModelResponse:
    # tool-mocking only applies to chat when we mocked pre-request
    if ctx.kind == "chat" and getattr(ctx, "raw_resp_needed", False):  # type: ignore[attr-defined]
        raw_resp = copy.deepcopy(resp)
        resp = self._post_response_prompt_mock(
            resp,
            nonfncall_msgs=getattr(ctx, "nonfn_msgs") or [],              # type: ignore[attr-defined]
            tools=ctx.tools,
        )
        # keep raw vs converted for telemetry parity
        self._telemetry.on_response(resp, raw_resp=raw_resp)
        return resp
    return resp
```

enyst commented on openhands/sdk/llm/llm.py:
```diff
@@ -493,6 +613,58 @@ def _normalize_call_kwargs(self, opts: dict, *, has_tools: bool) -> dict:
+
+        if "litellm_proxy" not in self.model:
+            out.pop("extra_body", None)
+        return out
```
I’ll make a pass with that approach!

enyst replied:
Done, cleaning up a bit!

xingyaoww commented on openhands/sdk/llm/llm.py:
```diff
@@ -493,6 +613,58 @@ def _normalize_call_kwargs(self, opts: dict, *, has_tools: bool) -> dict:
+            converted: list[dict] = []
+            for t in tools_param:
+                if hasattr(t, "to_responses"):
+                    converted.append(t.to_responses())  # type: ignore[attr-defined]
```
Ah interesting, can we modify .completion to accept list[Tool] and we do the conversion internally so we can simplify the branching here?

enyst commented:
I refactored the code with discriminator. Frankly, I don’t actually find it readable, I think because there is much more that needed separated and handled than it seemed at first; for example, the example in your comment above assumes that Responses API has no tools; adding them added quite a bit of code, variables, and checks, both in request and response.

(In general, I feel the approach is suitable if the similitudes are a lot more than differences 🤔; after the differences reach a threshold, I’m starting to think the entire conversion and/or approach has too high cognitive load effect - it makes the code less readable, or debuggable. But I’ll come back on structure in follow-up PRs. I want to propose another approach.)

Right now, the agent fixed a lot of tests without supervision, so I want to go through the code once more, it might need some cleanup - GPT-5 really likes its overly clever tricks. Will do around ~morning.

enyst commented on openhands/sdk/llm/utils/telemetry.py:
```diff
@@ -113,12 +113,20 @@ def _has_meaningful_usage(self, usage) -> bool:
 
     def _record_usage(
-        self, usage: Usage, response_id: str, context_window: int
+        self, usage: Usage, response_id: str, context_window: int | None
```
This was overdoing it; cleaned up!

enyst commented:
With the V1 CLI
<img width="760" height="406" alt="image" src="https://github.com/user-attachments/assets/3e9ddf06-c0a0-4ce8-93ba-bdc503d8bc78" />

enyst commented on openhands/sdk/llm/utils/responses_converter.py:
```diff
@@ -0,0 +1,219 @@
+            continue
+
+        # 3) Plain text messages
+        text = _to_text(msg.get("content", ""))
```
@openhands Fix @xingyaoww 's comment here! Note that we already support images, you should check how it works with Completions API and make sure to do similar for Responses API converter.

enyst replied:
This should be fixed!

xingyaoww commented on openhands/sdk/llm/utils/responses_converter.py:
```diff
@@ -0,0 +1,255 @@
+logger = get_logger(__name__)
+
+
+def messages_to_responses_items(messages: list[dict[str, Any]]) -> list[dict[str, Any]]:
```
not necessarily need in this pr - we should use more informative type annotation if possible: e.g., list[dict[str, Any]] should be list[ChatCompletionXXX] from litellm

xingyaoww replied:
actually i think maybe we should do it now -- just to make sure our conversion is robust and type-checked.. to save our future selfs from debugging llm issues 😭

xingyaoww commented on openhands/sdk/llm/llm.py:
```diff
@@ -295,6 +351,132 @@ def _set_env_side_effects(self):
+        if input is None:
+            if messages is not None:
+                # Allow a direct string to be used as input
+                if isinstance(messages, str):
```
this doesn't seem to match the input fucntion signature

xingyaoww commented on task.md:
```diff
@@ -0,0 +1,52 @@
+
```
We should remove this file

enyst commented on openhands/sdk/llm/llm.py:
```diff
@@ -295,6 +351,132 @@ def _set_env_side_effects(self):
+
+        if kind == "chat":
+            # Messages: ensure list[dict]
+            if messages and isinstance(messages[0], Message):
+                messages = self.format_messages_for_llm(cast(list[Message], messages))
+            else:
+                messages = cast(list[dict[str, Any]] | None, messages)
+
```
Yes! `input` can be str or a list of dicts compatible with `messages`

@openhands `messages` argument used to come from the client code as a list of dicts, but it no longer does: the client sends either a list of Messages or a string. Please remove the now-redundant `else`.

enyst commented:
@xingyaoww  The agent seems unable to find the LLM that I was testing with before:

> Result:
> - We can fetch model info from the proxy (it reports gpt-5-mini-2025-08-07, supports_reasoning=True), but the /responses call fails:
>   400 Bad Request: “There are no healthy deployments for litellm_proxy/openai/gpt-5-mini-2025-08-07”
> - That means the proxy knows about the model but doesn’t currently have a healthy deployment for it.
> 
> Where I “saw” the model
> - The proxy’s model-info endpoint returned data containing key: gpt-5-mini-2025-08-07. So it exists in metadata, but the runtime deployment isn’t healthy for /responses right now.

I'm going to try other variants!

enyst commented:
@xingyaoww  Thank you!

OK. I'm just looking also into simplifying `responses()`, since maybe `list[Message] | string` will do.

We also have the other option here:
- https://github.com/All-Hands-AI/agent-sdk/pull/396

You may want to see the .md file, I talked to the agent first to hash out the plan. 
(please ignore the code atm)

enyst commented:
@openhands Do a /coderoasted review of this PR. Make sure to include source/file/line when relevant.

enyst commented:
Heads-up: LiteLLM crashes before reaching the provider due to its logging deepcopy interacting with Pydantic-validated messages.

Context:
- Our request context is a Pydantic discriminated union (ReqCtx = Union[ChatCtx | ResponsesCtx], discriminator="kind"). For chat, ChatCtx(messages: list[AllMessageValues]) causes Pydantic v2 to validate LiteLLM's OpenAI message shapes.
- During validation of user.content (declared in LiteLLM types as Iterable[Union[ChatCompletionTextObject | ...]]), pydantic_core produces a ValidatorIterator. As a result, the structure stored in ChatCtx.messages for user.content is no longer a plain list.
- LiteLLM's logging layer (function_setup -> LiteLLMLogging.__init__) does copy.deepcopy(messages). deepcopy attempts to pickle that ValidatorIterator and throws:
  TypeError: cannot pickle 'pydantic_core._pydantic_core.ValidatorIterator' object

Net effect: the call fails inside LiteLLM pre-call logging (before any provider call), because Pydantic validation—triggered by our discriminated ChatCtx/messages typing—injects a pydantic_core iterator into user.content, and LiteLLM then deep-copies it.
